zk_ee/src/system/mod.rs

```diff
@@ -88,99 +90,95 @@ impl<S: SystemTypes> System<S> {
     }
 
     pub fn get_tx_origin(&self) -> <S::IOTypes as SystemIOTypesConfig>::Address {
-        self.metadata.tx_origin
+        self.metadata.tx_origin()
     }
 
     pub fn get_block_number(&self) -> u64 {
-        self.metadata.block_level_metadata.block_number
+        self.metadata.block_number()
     }
 
-    pub fn get_mix_hash(&self) -> ruint::aliases::U256 {
+    pub fn get_mix_hash(&self) -> Bytes32 {
         #[cfg(feature = "prevrandao")]
         {
-            self.metadata.block_level_metadata.mix_hash
+            self.metadata
+                .block_randomness()
+                .expect("block randomness should be provided")
         }
 
         #[cfg(not(feature = "prevrandao"))]
         {
-            ruint::aliases::U256::ONE
+            Bytes32::from_array(ruint::aliases::U256::ONE.to_be_bytes::<32>())
         }
     }
 
-    pub fn get_blockhash(&self, block_number: u64) -> ruint::aliases::U256 {
-        let current_block_number = self.metadata.block_level_metadata.block_number;
+    pub fn get_blockhash(&self, block_number: u64) -> Bytes32 {
+        let current_block_number = self.metadata.block_number();
         if block_number >= current_block_number
             || block_number < current_block_number.saturating_sub(256)
         {
             // Out of range
-            ruint::aliases::U256::ZERO
+            Bytes32::ZERO
         } else {
-            let index = 256 - (current_block_number - block_number);
-            self.metadata.block_level_metadata.block_hashes.0[index as usize]
+            let depth = current_block_number - block_number;
+            self.metadata
+                .block_historical_hash(depth)
+                .expect("historical hash of limited depth must be provided")
```

Comment:
Should we throw an internal error here? As well as in other places in this struct

Reply:
Sure, will do. It was a panic before, but better use internal error